File: src/lib/ring-buffer.ts
/** Fixed-capacity ring buffer. Oldest values are dropped; length never exceeds capacity. */
export class RingBuffer<T> {
  private readonly buf: Array<T | undefined>;
  private head = 0;
  private length_ = 0;

  constructor(readonly capacity: number) {
    if (capacity < 1) throw new Error("capacity must be >= 1");
    this.buf = new Array(capacity);
  }

  get length(): number {
    return this.length_;
  }

  push(item: T): void {
    const i = (this.head + this.length_) % this.capacity;
    this.buf[i] = item;
    if (this.length_ < this.capacity) {
      this.length_ += 1;
    } else {
      this.head = (this.head + 1) % this.capacity;
    }
  }

  toArray(): T[] {
    const out: T[] = [];
    for (let i = 0; i < this.length_; i++) {
      out.push(this.buf[(this.head + i) % this.capacity] as T);
    }
    return out;
  }
}

File: backend/app/services/ring_buffer.rb
# frozen_string_literal: true

class RingBuffer
  attr_reader :capacity, :length

  def initialize(capacity)
    raise ArgumentError, "capacity must be >= 1" if capacity < 1

    @capacity = capacity
    @buf = Array.new(capacity)
    @head = 0
    @length = 0
  end

  def push(item)
    i = (@head + @length) % @capacity
    @buf[i] = item
    if @length < @capacity
      @length += 1
    else
      @head = (@head + 1) % @capacity
    end
    self
  end

  def to_a
    Array.new(@length) { |i| @buf[(@head + i) % @capacity] }
  end
end
